src/Headers/Ic.h

```diff
@@ -0,0 +1,984 @@
+/// \details Class containing functions for generating initial conditions.
+/// \author  D. A. Hubber, G. Rosotti
+/// \date    02/02/2015
+//=================================================================================================
+template <int ndim>
+class Ic
+{
+protected:
+
+  Simulation<ndim>* const sim;              ///< Simulation class pointer
+  Hydrodynamics<ndim>* const hydro;         ///< Hydrodynamics algorithm pointer
+  const FLOAT invndim;                      ///< 1/ndim
+  const SimUnits& simunits;                 ///< Reference to main simunits object
+  const DomainBox<ndim>& simbox;            ///< Reference to simulation bounding box object
+
+  int Ntable;                               ///< No. of table elements
+  FLOAT *xTable;                            ///< Tabulated position values
+  FLOAT *mTable;                            ///< Tabulated integrated mass values
+  FLOAT *mFracTable;                        ///< Tabulated fractional integrated mass values
+  std::string posQuantity;                  ///< Position quantity string (e.g. x, y, r)
+  Parameters* simparams;                    ///< Pointer to parameters object
+  RandomNumber *randnumb;                   ///< Random number object pointer
+
+
+  // Helper routines
+  //-----------------------------------------------------------------------------------------------
+  void AddAzimuthalDensityPerturbation(const int, const int, const FLOAT, const FLOAT *, FLOAT *);
+  void AddBinaryStar(FLOAT, FLOAT, FLOAT, FLOAT, FLOAT, FLOAT, FLOAT, FLOAT, FLOAT, FLOAT,
+                     FLOAT *, FLOAT *, NbodyParticle<ndim> &, NbodyParticle<ndim> &);
+  void AddRotationalVelocityField(const int, const FLOAT, const FLOAT *, const FLOAT *, FLOAT *);
+  void Addr2Sphere(int, FLOAT *, FLOAT *, FLOAT);
+  void AddSinusoidalDensityPerturbation(int, FLOAT, FLOAT, FLOAT *);
+  void GenerateTurbulentVelocityField(const int, const int, const DOUBLE, DOUBLE *);
+  void InterpolateVelocityField(const int, const int, const FLOAT, const FLOAT,
+                                const FLOAT *, const DOUBLE *, FLOAT *);
+
+  void AddMonteCarloDensityField(const int, const Box<ndim> &, FLOAT *, RandomNumber *);
+  virtual FLOAT GetMaximumDensity(const Box<ndim> &, RandomNumber *);
+
+
+public:
+
+  // Constructor for setting important variables and pointers
+  //-----------------------------------------------------------------------------------------------
+  Ic(Simulation<ndim>* _sim, Hydrodynamics<ndim>* _hydro, FLOAT _invndim) :
+    sim(_sim), hydro(_hydro), invndim(_invndim),
+    simunits(_sim->simunits), simbox(_sim->simbox),
+    simparams(_sim->simparams), randnumb(_sim->randnumb)
+  {
+  };
+  virtual ~Ic() {};
+
+
+  // Virtual functions
+  //-----------------------------------------------------------------------------------------------
+  virtual void Generate(void) {};
+  virtual FLOAT GetValue(const std::string, const FLOAT *) {return (FLOAT) 0.0;}
```

Comment:
GetValue simply returns the (floating point) value of some quantity (with the given string name) at the given position, for the purpose of defining all IC quantities in one place.  It's hopefully a straightforward function that will cover 99% of ICs that anyone can write.